### PR TITLE
Show PowerShell cmdlets in output when running in PowerShell

### PR DESCRIPTION
**Pull request by beechtom**
When running certain commands or raising certain errors, Bolt will
output a command to run. This change checks whether Bolt is running in
PowerShell or not, and sets the command in the output to a PowerShell
cmdlet if it is or a *nix shell command if it is not.

!no-release-note

### CODE REVIEW

File: lib/bolt/cli.rb [hunk 1]
```diff
@@ -229,9 +229,9 @@ def update_targets(options)
 
     def validate(options)
       unless COMMANDS.include?(options[:subcommand])
+        command = Bolt::Util.powershell? ? 'Get-Command -Module PuppetBolt' : 'bolt help'
```
Comment (lucywyman) [inline lib/bolt/cli.rb:232]: This doesn't work if the user hasn't loaded the powershell module. I think we can see if the user has the module loaded with `Get-Module -ListAvailable -Name PuppetBolt` and switch on that instead?

File: lib/bolt/cli.rb [hunk 2]
```diff
@@ -229,9 +229,9 @@ def update_targets(options)
...
         raise Bolt::CLIError,
-              "Expected subcommand '#{options[:subcommand]}' to be one of " \
-              "#{COMMANDS.keys.join(', ')}"
+              "'#{options[:subcommand]}' is not a Bolt command. See '#{command}'."
```
Comment (lucywyman) [inline lib/bolt/cli.rb:234]: Dang much better phrasing :clap:

**Review by lucywyman — COMMENTED** [review]
I think a few places were also missed:
https://github.com/puppetlabs/bolt/blob/main/lib/bolt/bolt_option_parser.rb#L199
https://github.com/puppetlabs/bolt/blob/main/lib/bolt/project_migrator/config.rb#L60
https://github.com/puppetlabs/bolt/blob/main/lib/bolt/outputter/human.rb#L299
https://github.com/puppetlabs/bolt/blob/main/lib/bolt/inventory/group.rb#L247
And in cli.rb at lines:
290
469
869
872
and 875

File: lib/bolt/cli.rb (hunk 1 above)
Comment (beechtom) [reply lib/bolt/cli.rb:232]: Shouldn't the powershell module be loaded automatically when Bolt is installed as a package?

File: lib/bolt/cli.rb (hunk 2 above)
Comment (beechtom) [reply lib/bolt/cli.rb:234]: Thank you git for the inspo ;)

**Comment by beechtom** [comment]
Updated.

I left out https://github.com/puppetlabs/bolt/blob/main/lib/bolt/bolt_option_parser.rb#L199 because that help text isn't used in the Powershell module and is already specific to the *nix shell commands. It seems incorrect to show a Powershell cmdlet in the help text for *nix shell commands.

Similarly, https://github.com/puppetlabs/bolt/blob/main/lib/bolt/outputter/human.rb#L299 was left out since there is no Powershell cmdlet for the topic guides and this list won't be shown when using the Powershell cmdlet `Get-Help about_bolt_*`.

File: lib/bolt/cli.rb (hunk 1 above)
Comment (lucywyman) [reply lib/bolt/cli.rb:232]: https://puppet.com/docs/bolt/latest/bolt_installing.html#add-the-bolt-module-to-powershell I doubt many people are using Powershell 2 or 3 but it generally just seems better to verify the cmdlets are loaded before recommending a cmdlet? Does `Get-Module` add a lot of time? There might be more efficient ways to check...

File: lib/bolt/cli.rb [hunk 3]
```diff
@@ -861,17 +866,35 @@ def install_puppetfile(config, puppetfile, moduledir)
...
+        when 'generate-types'
+          old_command = 'Register-BoltPuppetfileTypes'
+          new_command = 'Register-BoltModuleTypes'
+        when 'install'
+          old_command = 'Install-BoltPuppetfile'
+          new_command = 'Install-BoltModule'
+        when 'show-modules'
+          old_command = 'Get-BoltPuppetfileModules'
+          new_command = 'Get-BoltModule'
+        end
```
Comment (lucywyman) [inline lib/bolt/cli.rb:880]: This also needs a clause for 'add' and 'show' if you're using the `bolt module` command without `modules` in config.

File: lib/bolt/cli.rb [hunk 4]
```diff
@@ -837,11 +840,13 @@ def add_project_module(name, project)
     def assert_project_file(project)
```
Comment (lucywyman) [inline lib/bolt/cli.rb:840]: Huh! This is actually dead code, because if the user doesn't have a projectfile this will be raised:
```
Unable to use command 'bolt module install' when 'modules' is not configured in bolt-project.yaml. Use 'bolt puppetfile install' instead.
```
From line 202 first. I think I'd move `assert_puppfile_or_module` to be called in the `puppetfile` and `module` handlers, maybe even just after `assert_projectfile`, but you might have better ideas of how that flow should work.

File: lib/bolt/module_installer.rb [hunk 5]
```diff
@@ -194,5 +194,29 @@ def install_puppetfile(path, moduledir, config = {})
 
       ok
     end
+
+    # Asserts that an existing Puppetfile is managed by Bolt.
+    #
+    private def assert_managed_puppetfile(puppetfile, path)
```
Comment (lucywyman) [inline lib/bolt/module_installer.rb:200]: I don't see where this gets called?

File: lib/bolt/outputter/human.rb [hunk 6]
```diff
@@ -225,20 +226,26 @@ def print_task_info(task)
...
+                else
+                  +"bolt task run #{task.name} --targets <targets>"
+                end
 
         task.parameters&.each do |k, v|
           pretty_params << "- #{k}: #{v['type'] || 'Any'}\n"
           pretty_params << "    Default: #{v['default'].inspect}\n" if v.key?('default')
           pretty_params << "    #{v['description']}\n" if v['description']
-          usage << if v['type'].is_a?(Puppet::Pops::Types::POptionalType)
+          usage << if v['type'].start_with?("Optional")
```
Comment (lucywyman) [inline lib/bolt/outputter/human.rb:239]: Why this change?

File: lib/bolt/util.rb [hunk 7]
```diff
@@ -273,6 +273,11 @@ def windows?
         !!File::ALT_SEPARATOR
       end
 
+      # Returns true if running in PowerShell.
+      def powershell?
+        !!ENV['PSModulePath']
```
Comment (lucywyman) [inline lib/bolt/util.rb:278]: @jpogran Does this seem like a reasonable way to know if we're in powershell vs. CMD or *sh?

Comment (beechtom) [reply lib/bolt/util.rb:278]: This is actually his suggestion!

File: lib/bolt/outputter/human.rb (hunk 6 above)
Comment (beechtom) [reply lib/bolt/outputter/human.rb:239]: Because it's a bug. `v['type']` is the stringified type, so `is_a?(Puppet::Pops::Types::POptionalType)` will always be false.

File: lib/bolt/module_installer.rb (hunk 5 above)
Comment (beechtom) [reply lib/bolt/module_installer.rb:200]: Leftover from before the refactor of this class.

File: lib/bolt/cli.rb (hunk 1 above)
Comment (lucywyman) [reply lib/bolt/cli.rb:228]: Ok, it looks like in the last month, out of 278 Windows users we've had 7 on Windows 7 and Windows 2012 (non-r2), which have Powershell3 by default. I feel ok leaving this as is and modifying if anyone complains.
